### PR TITLE
Liu_LA119.solar_china

### PR DESCRIPTION
**Pull request by yanliuthu**
Add zchunk_LA119.solar_china to gcamchina

### CODE REVIEW

File: R/zchunk_LA119.solar_china.R [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+    all_data <- list(...)[[1]]
+    # -----------------------------------------------------------------------------
+    # 1.Load required inputs
+    province_names_mappings     <- get_data(all_data, "gcam-china/province_names_mappings")
+    solar_csp_pv_capacityfactor <- get_data(all_data, "gcam-china/solar_csp_pv_capacityfactor")
+
+    # -----------------------------------------------------------------------------
+    # 2.perform computations
+    filter(solar_csp_pv_capacityfactor, province.name == "Average")$Centralized.PV -> PV_ave
```
Comment (rynacui) [inline R/zchunk_LA119.solar_china.R]: I think these two lines are not necessary, can be combined into the following pipe.  See below

File: R/zchunk_LA119.solar_china.R [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    filter(solar_csp_pv_capacityfactor, province.name == "Average")$CSP -> CSP_ave
+
+    solar_csp_pv_capacityfactor %>%
+      mutate(sector = "electricity generation",
+             fuel = "solar PV",
+             scaler = Centralized.PV / PV_ave) %>%
```
Comment (rynacui) [inline R/zchunk_LA119.solar_china.R]: `scaler = Centralized.PV / Centralized.PV[province.name == "Average"]`

File: R/zchunk_LA119.solar_china.R [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+             scaler = CSP / CSP_ave) %>%
+      filter(province.name != "Average") %>%
+      map_province_name(province_names_mappings, "province", TRUE) %>%
+      select(province, sector, fuel, scaler) ->
+      L119.CapFacScaler_CSP_province
+
+
+    # ===================================================
+    L119.CapFacScaler_PV_province %>%
+      add_title("ScalarSolar PV capacity factor adjustment by province") %>%
```
Comment (rynacui) [inline R/zchunk_LA119.solar_china.R]: Is `Scalar` necessary?

File: R/zchunk_LA119.solar_china.R [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+#'
+#' Compute capacity factors for wind by province.
+#'
+#' @param command API command to execute
+#' @param ... other optional parameters, depending on command
+#' @return Depends on \code{command}: either a vector of required inputs,
+#' a vector of output names, or (if \code{command} is "MAKE") all
+#' the generated outputs: \code{L119.CapFacScaler_PV_province}, \code{L119.CapFacScaler_CSP_province}. The corresponding file in the
+#' original data system was \code{LA114.Solar.R} (gcam-china level1).
+#' @details This chunk computes scalars by province to vary capacity factors for central station PV and CSP technologies by province.
```
Comment (rynacui) [inline R/zchunk_LA119.solar_china.R]: sth like: "computes solar capacity factor scalars for PV and CSP technologies by province, divided by the national average"

File: R/zchunk_LA119.solar_china.R [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+module_gcam.china_LA119.Solar <- function(command, ...) {
+  if(command == driver.DECLARE_INPUTS) {
+    return(c(FILE = "gcam-china/solar_csp_pv_capacityfactor",
+             FILE = "gcam-china/province_names_mappings"))
+  } else if(command == driver.DECLARE_OUTPUTS) {
+    return(c("L119.CapFacScaler_PV_province",
+             "L119.CapFacScaler_CSP_province"))
+  } else if(command == driver.MAKE) {
+
+    fuel  <- CSP <- .  <- sector <- scaler <-
```
Comment (rynacui) [inline R/zchunk_LA119.solar_china.R]: `.` is not needed

File: R/zchunk_LA119.solar_china.R [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+      map_province_name(province_names_mappings, "province", TRUE) %>%
+      select(province, sector, fuel, scaler) ->
+      L119.CapFacScaler_CSP_province
+
+
+    # ===================================================
+    L119.CapFacScaler_PV_province %>%
+      add_title("ScalarSolar PV capacity factor adjustment by province") %>%
+      add_units("Unitless") %>%
+      add_comments("The scalars are generated by dividing data on capacity factors by province by national average capacity factor from rNREL") %>%
```
Comment (rynacui) [inline R/zchunk_LA119.solar_china.R]: `NREL`?

File: R/zchunk_LA119.solar_china.R [hunk 7]
```diff
@@ -0,0 +1,79 @@
+#' module_gcam.china_LA119.Solar
+#'
+#' Compute capacity factors for wind by province.
```
[the remark this answers is not in the document]
Comment (yanliuthu) [reply R/zchunk_LA119.solar_china.R]: Have been changed.

File: R/zchunk_LA119.solar_china.R (hunk 2 above)
Comment (yanliuthu) [reply R/zchunk_LA119.solar_china.R]: Have been changed.

File: R/zchunk_LA119.solar_china.R (hunk 3 above)
Comment (yanliuthu) [reply R/zchunk_LA119.solar_china.R]: No, it has been changed.

File: R/zchunk_LA119.solar_china.R (hunk 7 above)
[the remark this answers is not in the document]
Comment (yanliuthu) [reply R/zchunk_LA119.solar_china.R]: Has been changed.

File: R/zchunk_LA119.solar_china.R (hunk 2 above)
Comment (yanliuthu) [reply R/zchunk_LA119.solar_china.R]: Have been changed.

File: R/zchunk_LA119.solar_china.R (hunk 4 above)
Comment (yanliuthu) [reply R/zchunk_LA119.solar_china.R]: OK

File: R/zchunk_LA119.solar_china.R (hunk 6 above)
Comment (yanliuthu) [reply R/zchunk_LA119.solar_china.R]: I'm not sure where the national average comes from (NREL is the source of gcam-usa), so I just remove it.

Comment (rynacui) [reply R/zchunk_LA119.solar_china.R]: Please do not copy the source from gcam-usa. Check with @sypnl about data source. Also for the other output.

File: inst/extdata/gcam-china/solar_csp_pv_capacityfactor.csv [hunk 8]
```diff
@@ -0,0 +1,38 @@
+# File: solar_csp_pv_capacityfactor.csv
+# Title: Capacity factor for centralized PV and CSP by province
+# Units: nan
```
Comment (sypnl) [inline inst/extdata/gcam-china/solar_csp_pv_capacityfactor.csv]: # Units: Unitless

File: R/zchunk_LA119.solar_china.R [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+      province.nam <- province <- Centralized.PV <- NULL     # silence package check.
+
+    all_data <- list(...)[[1]]
+    # -----------------------------------------------------------------------------
+    # 1.Load required inputs
+    province_names_mappings     <- get_data(all_data, "gcam-china/province_names_mappings")
+    solar_csp_pv_capacityfactor <- get_data(all_data, "gcam-china/solar_csp_pv_capacityfactor")
+
+    # -----------------------------------------------------------------------------
+    # 2.perform computations
```
Comment (sypnl) [inline R/zchunk_LA119.solar_china.R]: # Create scalers to scale capacity factors read in the assumptions file.
# These scalers will then be used to create capacity factors by province.
# The idea is to vary capacity factors for solar technologies by province depending on the varying solar irradiance by province.

 # Create scalers by province by dividing capacity factor by the average

**Pull request merged**